azurelinuxagent/ga/self_update_version_updater.py

```diff
@@ -61,34 +63,36 @@ def _get_agent_upgrade_type(version):
         return SelfUpdateType.Regular
 
     @staticmethod
-    def _get_next_process_time(last_val, frequency, now):
+    def _get_next_process_time(upgrade_type, now):
         """
-        Get the next upgrade time
+        Returns random time in between 0 to 24hrs(regular) or 6hrs(hotfix) from now
         """
-        return now if last_val == datetime.datetime.min else last_val + datetime.timedelta(seconds=frequency)
+        if upgrade_type == SelfUpdateType.Hotfix:
+            frequency = conf.get_self_update_hotfix_frequency()
+        else:
+            frequency = conf.get_self_update_regular_frequency()
+        return now + datetime.timedelta(seconds=random.randint(0, frequency))
 
-    def _is_new_agent_allowed_update(self):
+    def _new_agent_allowed_now_to_update(self):
         """
-        This method ensure that update is allowed only once per (hotfix/Regular) upgrade frequency
+        This method called when new update detected and computes random time for next update.
+        If the current time on or after upgrade time, we allow the update.
+
+        Note: After we allow the update, and it's not successful, the next update time will be recalculated.
         """
         now = datetime.datetime.utcnow()
         upgrade_type = self._get_agent_upgrade_type(self._version)
-        if upgrade_type == SelfUpdateType.Hotfix:
-            next_update_time = self._get_next_process_time(self._last_attempted_self_update_time,
-                                                           conf.get_self_update_hotfix_frequency(), now)
-        else:
-            next_update_time = self._get_next_process_time(self._last_attempted_self_update_time,
-                                                           conf.get_self_update_regular_frequency(), now)
 
-        if self._version > CURRENT_VERSION:
-            message = "Self-update discovered new {0} upgrade WALinuxAgent-{1}; Will upgrade on or after {2}".format(
-                upgrade_type, str(self._version), next_update_time.strftime(logger.Logger.LogTimeFormatInUTC))
-            logger.info(message)
-            add_event(op=WALAEventOperation.AgentUpgrade, message=message, log_event=False)
-
-        if next_update_time <= now:
-            # Update the last upgrade check time even if no new agent is available for upgrade
-            self._last_attempted_self_update_time = now
+        if not self._update_time_refreshed:
```

Comment:
could you add some comments about the intent of self._update_time_refreshed? thanks

Reply:
actually, after review, I realized I could achieve same thing without _update_time_refreshed flag. Added more details in the code comments